### ISSUE TITLE
AttributeError: 'FSMap' object has no attribute 'missing_exceptions'

### ISSUE THREAD

**Issue by chiaral** [dask/gcsfs#299]
Apologies for not creating a simpler reproducible example. Unfortunately, I cannot reproduce the error with simple examples, only with a more realistic example involving real data and code. I hope that by sharing it here, the error message will trigger some ideas about how to resolve it.

**What happened**:

I am trying to generate a large xarray dataset using Dask and write to to GCS using zarr + gcsfs with a distributed dask gateway cluster.

My code looks something like this

```python
import gcsfs
import json

with open('access_token.json') as f:
    token = json.load(f)
    gcs = gcsfs.GCSFileSystem(project='pangeo-181919', token=token)

# now do a bunch of lazy computations to build up an xarray dataset called ds_to_save
# I know these computations are not the problem because my dataset can be computed fine

gcsmap = gcs.get_mapper('/pangeo-cmip6-cape/CanESM5/historical/Convective_'+np.str(year)+'.zarr')
ds_to_save.to_zarr(gcsmap, mode='w')
```

This produces a long stack trace

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-13-35e977e44506> in <module>
    107     #gcsmap = gcsfs.GCSMap('/pangeo-cmip6-cape/CanESM5/historical/Convective_'+np.str(year)+'.zarr', gcs=gcs)
    108     gcsmap = gcs.get_mapper('/pangeo-cmip6-cape/CanESM5/historical/Convective_'+np.str(year)+'.zarr')
--> 109     ds_to_save.to_zarr(gcsmap, mode='w')
    110     end = time.time()
    111     print(end - start)

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/dataset.py in to_zarr(self, store, chunk_store, mode, synchronizer, group, encoding, compute, consolidated, append_dim)
   1660             compute=compute,
   1661             consolidated=consolidated,
-> 1662             append_dim=append_dim,
   1663         )
   1664 

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/api.py in to_zarr(dataset, store, chunk_store, mode, synchronizer, group, encoding, compute, consolidated, append_dim)
   1367     # TODO: figure out how to properly handle unlimited_dims
   1368     dump_to_store(dataset, zstore, writer, encoding=encoding)
-> 1369     writes = writer.sync(compute=compute)
   1370 
   1371     if compute:

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/common.py in sync(self, compute)
    159                 compute=compute,
    160                 flush=True,
--> 161                 regions=self.regions,
    162             )
    163             self.sources = []

/srv/conda/envs/notebook/lib/python3.7/site-packages/dask/array/core.py in store(sources, targets, lock, regions, compute, return_stored, **kwargs)
    979 
    980         if compute:
--> 981             result.compute(**kwargs)
    982             return None
    983         else:

/srv/conda/envs/notebook/lib/python3.7/site-packages/dask/base.py in compute(self, **kwargs)
    165         dask.base.compute
    166         """
--> 167         (result,) = compute(self, traverse=False, **kwargs)
    168         return result
    169 

/srv/conda/envs/notebook/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    450         postcomputes.append(x.__dask_postcompute__())
    451 
--> 452     results = schedule(dsk, keys, **kwargs)
    453     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    454 

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in get(self, dsk, keys, restrictions, loose_restrictions, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2723                     should_rejoin = False
   2724             try:
-> 2725                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   2726             finally:
   2727                 for f in futures.values():

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in gather(self, futures, errors, direct, asynchronous)
   1990                 direct=direct,
   1991                 local_worker=local_worker,
-> 1992                 asynchronous=asynchronous,
   1993             )
   1994 

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    831         else:
    832             return sync(
--> 833                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    834             )
    835 

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    338     if error[0]:
    339         typ, exc, tb = error[0]
--> 340         raise exc.with_traceback(tb)
    341     else:
    342         return result[0]

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/utils.py in f()
    322             if callback_timeout is not None:
    323                 future = asyncio.wait_for(future, callback_timeout)
--> 324             result[0] = yield future
    325         except Exception as exc:
    326             error[0] = sys.exc_info()

/srv/conda/envs/notebook/lib/python3.7/site-packages/tornado/gen.py in run(self)
    733 
    734                     try:
--> 735                         value = future.result()
    736                     except Exception:
    737                         exc_info = sys.exc_info()

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1849                             exc = CancelledError(key)
   1850                         else:
-> 1851                             raise exception.with_traceback(traceback)
   1852                         raise exc
   1853                     if errors == "skip":

/srv/conda/envs/notebook/lib/python3.7/site-packages/dask/array/core.py in store_chunk()
   3722 
   3723 def store_chunk(x, out, index, lock, return_stored):
-> 3724     return load_store_chunk(x, out, index, lock, return_stored, False)
   3725 
   3726 

/srv/conda/envs/notebook/lib/python3.7/site-packages/dask/array/core.py in load_store_chunk()
   3711     try:
   3712         if x is not None:
-> 3713             out[index] = np.asanyarray(x)
   3714         if return_stored and load_stored:
   3715             result = out[index]

/srv/conda/envs/notebook/lib/python3.7/site-packages/zarr/core.py in __setitem__()
   1113 
   1114         fields, selection = pop_fields(selection)
-> 1115         self.set_basic_selection(selection, value, fields=fields)
   1116 
   1117     def set_basic_selection(self, selection, value, fields=None):

/srv/conda/envs/notebook/lib/python3.7/site-packages/zarr/core.py in set_basic_selection()
   1208             return self._set_basic_selection_zd(selection, value, fields=fields)
   1209         else:
-> 1210             return self._set_basic_selection_nd(selection, value, fields=fields)
   1211 
   1212     def set_orthogonal_selection(self, selection, value, fields=None):

/srv/conda/envs/notebook/lib/python3.7/site-packages/zarr/core.py in _set_basic_selection_nd()
   1499         indexer = BasicIndexer(selection, self)
   1500 
-> 1501         self._set_selection(indexer, value, fields=fields)
   1502 
   1503     def _set_selection(self, indexer, value, fields=None):

/srv/conda/envs/notebook/lib/python3.7/site-packages/zarr/core.py in _set_selection()
   1548 
   1549             # put data
-> 1550             self._chunk_setitem(chunk_coords, chunk_selection, chunk_value, fields=fields)
   1551 
   1552     def _chunk_getitem(self, chunk_coords, chunk_selection, out, out_selection,

/srv/conda/envs/notebook/lib/python3.7/site-packages/zarr/core.py in _chunk_setitem()
   1663         with lock:
   1664             self._chunk_setitem_nosync(chunk_coords, chunk_selection, value,
-> 1665                                        fields=fields)
   1666 
   1667     def _chunk_setitem_nosync(self, chunk_coords, chunk_selection, value, fields=None):

/srv/conda/envs/notebook/lib/python3.7/site-packages/zarr/core.py in _chunk_setitem_nosync()
   1693 
   1694                 # obtain compressed data for chunk
-> 1695                 cdata = self.chunk_store[ckey]
   1696 
   1697             except KeyError:

/srv/conda/envs/notebook/lib/python3.7/site-packages/fsspec/mapping.py in __getitem__()
    131         try:
    132             result = self.fs.cat(k)
--> 133         except self.missing_exceptions:
    134             if default is not None:
    135                 return default

AttributeError: 'FSMap' object has no attribute 'missing_exceptions'
```
By monitoring the dashboard, I can see that this error comes at the very end of a long, successful computation.

I can verify that the `gcsmap` object DOES have these attributes:
```python
print(gcsmap.missing_exceptions)
# -> (FileNotFoundError, IsADirectoryError, NotADirectoryError)
``` 

So this feels like a serialization problem. Also an internal problem with fsspec.

I would appreciate any advice you may have.

**Environment**:

This is in the latest pangeo-notebook environment.

- Fsspec version: 0.7.1
- Dask version: 2.30.0
- Python version: 3.7.8 
- Operating System: linux
- Install method (conda, pip, source): conda

**Comment by chiaral** [comment]
Also worth noting that this exact workflow used to work fine. So possibly related to some recent changes?

**Comment by martindurant** [comment]
I believe this is fixed in fsspec 0.8.4 https://github.com/intake/filesystem_spec/commit/8fc099dac4f4427f2eb8971e8efc7ede8756a7b2
